### PR TITLE
Use ng-router-loader

### PR DESCRIPTION
**Pull request by shlomiassaf**
This PR replaces the `angular-router-loader` with the `ng-router-loader`

For more details read [here](https://github.com/shlomiassaf/ng-router-loader)

@gdi2290

### CODE REVIEW

**Comment by colinskow** [comment]
Looks good to me. It would be nice if you had a test that fails with `angular-router-loader` but passes with `ng-router-loader` to show the difference.

My main question is how committed you are to keeping your loader up-to-date as Angular evolves?

**Comment by d3viant0ne** [comment]
I'm all for supporting emerging libs & @shlomiassaf has other libs that are well maintained, his modal lib being the longest running iirc.

As far as this particular library, I'm going to carve out some time in the morning to add it to a branch of an existing production application before actually reviewing. I like the idea and this certainly does solve an existing problem but It's one thing to be of use in a starter and something else entirely in a large production application.

In short, I'm okay with this pending how it performs in a non-trivial setting.

**Comment by shlomiassaf** [comment]
It's running in a production app at work.

I only built it because the angular-router-loader isn't really webpack friendly...

It just don't work with AOT...

Just look at the genDir defined in common, totally out of sync with the genDir in tscofig.

compiled/src/app VS compiled

Since it works you can rest assure something else won't once you get to different directory levels.

This is just the smallest problem... 

Im fine with reviewing it, I actually prefer to get feedback...  but angular-router-loader is far behind and doesn't really work in real world.. so we better of with this one

**Comment by JonnyBGod** [comment]
It seems to load only a few lazy loaded modules.

app.routes.ts
```
...
export const ROUTES: Routes = [
  {
    path: '',
    loadChildren: './components/index/index.module#IndexModule'
  },
  {
    path: 'user',
    loadChildren: './components/user/user.module#UserModule'
  },
  { path: 'login', redirectTo: 'user/login', pathMatch: 'full' },
  { path: 'signin', redirectTo: 'user/login', pathMatch: 'full' },
  { path: 'signup', redirectTo: 'user/signup', pathMatch: 'full' },
  {
    path: 'verify/:token/:userId',
    redirectTo: 'user/verify/:token/:userId',
    pathMatch: 'full'
  },
  { path: 'verify', redirectTo: 'user/verify', pathMatch: 'full' },
  {
    path: 'reset-password/:token/:userId',
    redirectTo: 'user/reset-password/:token/:userId',
    pathMatch: 'full'
  },
  { path: 'recover-account', redirectTo: 'user/recover-account', pathMatch: 'full' },

  // Private
  {
    path: 'apps',
    loadChildren: './components/apps/apps.module#AppsModule',
    canActivate: [ UserLoggedGuard ]
  },
  {
    path: 'settings',
    loadChildren: './components/settings/settings.module#SettingsModule',
    canActivate: [ UserLoggedGuard ]
  },

  // Public
  {
    path: ':id',
    loadChildren: './components/profile/profile.module#ProfileModule',
    resolve: {
      profile: ProfileResolver
    }
  },
  {
    path: '**',
    component: NotFoundComponent
  }
];
```

debug
```
...
webpack: bundle is now INVALID.
 10% building modules 6/7 modules 1 active .../seeds/web-seed/src/app/app.routes.ts================================== ng-router-loader ==================================
Importer:    ...web-seed/src/app/app.routes.ts
Raw Request: loadChildren: './components/index/index.module#IndexModule'
Replacement: loadChildren: () => new Promise(function (resolve) {  (require as any).ensure([], function (require: any) {    resolve(require('...web-seed/src/app/components/index/index.module')['IndexModule']);  });})
======================================================================================
 10% building modules 8/10 modules 2 active ...app/components/index/index.routing.ts================================== ng-router-loader ==================================
Importer:    ...web-seed/src/app/components/index/index.routing.ts
Raw Request: loadChildren: './home/home.module#IndexHomeModule'
Replacement: loadChildren: () => new Promise(function (resolve) {  (require as any).ensure([], function (require: any) {    resolve(require('...web-seed/src/app/components/index/home/home.module')['IndexHomeModule']);  });})
======================================================================================
633ms building modules
59ms sealing
0ms optimizing
0ms basic module optimization
293ms module optimization
...
```

Same config works with current master.
Current master does not work win AOT + nested lazy loaded modules.

**Comment by shlomiassaf** [comment]
@JonnyBGod Thanks for the feedback.

Do you have a repo with that project so I can test on?

**Comment by shlomiassaf** [comment]
@JonnyBGod Found the issue, greedy regex... will fix and udpate

**Comment by shlomiassaf** [comment]
@JonnyBGod Please update to `ng-router-loader@1.0.1` in your project, it should solve it.

Please update.

Thanks!

**Comment by shlomiassaf** [comment]
@d3viant0ne @gdi2290 @colinskow I'v added some nested & multiple router samples ...

The `+detail` module loads async and then has an aux route to another async loaded module (`+child-detail`)

Also added the `barrel` module to show multiple async routes working...
The `+barrel` module loads async and then has an aux route to another async loaded module (`+barrel-detail`)

The barrel might be too much, take a loot.

**Comment by gdi2290** [comment]
LGTM

**Comment by JonnyBGod** [comment]
Now it is working great. Tested with 30+ lazy loaded nested modules.

**Comment by shlomiassaf** [comment]
@JonnyBGod did it work?

**Comment by JonnyBGod** [comment]
Yes sorry forgot to mentions AOT working fine as well with both async loaders:
- "async-system"
- "async-require"

**Comment by shlomiassaf** [comment]
Thank you for the help!

**Comment by JonnyBGod** [comment]
I do not own a windows machine but a friend is reporting to me that it is not working on windows.

```
client:101./src/app/app.routes.ts
Module not found: Error: Can't resolve 'C:UsersjonatDesktopangular2-webpack-advance-startersrcappcomponentsaboutabout.module' in 'C:\Users\jonat\Desktop\angular2-webpack-advance-starter\src\app'
 @ ./src/app/app.routes.ts 5:43-153
 @ ./src/app/app.routerModule.ts
 @ ./src/app/app.advance.ts
 @ ./src/app/app.module.ts
 @ ./src/app/index.ts
 @ ./src/main.browser.ts
 @ multi main
```

**Comment by d3viant0ne** [comment]
Anyone in this thread have access to a windows machine to verify this functionality on Windoz ?

**Comment by shlomiassaf** [comment]
Please do not merge until I verify this, thank.

**Comment by shlomiassaf** [comment]
Fixed now, tested on Win7, Node 6.9.3 

```
shlassaf@SHLASSAF-32A69 MINGW64 ~/Desktop/CODE/angular2-webpack-starter (feat/ng-router-loader)
$ npm run start

> angular2-webpack-starter@5.4.1 start C:\Users\shlassaf\Desktop\CODE\angular2-webpack-starter
> npm run server:dev


> angular2-webpack-starter@5.4.1 server:dev C:\Users\shlassaf\Desktop\CODE\angular2-webpack-starter
> webpack-dev-server --config config/webpack.dev.js --progress --profile --watch --content-base src/

 10% building modules 2/3 modules 1 active ...client\index.js?http://localhost:3000
Project is running at http://localhost:3000/
webpack output is served from /
Content not from webpack is served from C:\Users\shlassaf\Desktop\CODE\angular2-webpack-starter\src
404s will fallback to /index.html                                                                                                                                                            10% building modules 4/9 modules 5 active ...r\node_modules\webpack\hot\emitter.js
[at-loader] Using typescript@2.0.10 from typescript and "tsconfig.json" from C:\Users\shlassaf\Desktop\CODE\angular2-webpack-starter\tsconfig.webpack.json (in a forked process).
                                                                                                                                                                                             63% building modules 445/478 modules 33 active ...r\core\src\linker\element_injector.js================================== ng-router-loader ==================================
Importer:    C:\Users\shlassaf\Desktop\CODE\angular2-webpack-starter\src\app\app.routes.ts
Raw Request: loadChildren: './+detail#DetailModule'
Replacement: loadChildren: () => System.import('C:\\Users\\shlassaf\\Desktop\\CODE\\angular2-webpack-starter\\src\\app\\+detail\\index')  .then( (module: any) => module['DetailModule'] )
======================================================================================
================================== ng-router-loader ==================================
Importer:    C:\Users\shlassaf\Desktop\CODE\angular2-webpack-starter\src\app\app.routes.ts
Raw Request: loadChildren: './+barrel#BarrelModule'
Replacement: loadChildren: () => System.import('C:\\Users\\shlassaf\\Desktop\\CODE\\angular2-webpack-starter\\src\\app\\+barrel\\index')  .then( (module: any) => module['BarrelModule'] )
======================================================================================                                                                                                  69% building modules 787/791 modules 4 active ...lar\common\src\private_import_core.js================================== ng-router-loader ==================================
Importer:    C:\Users\shlassaf\Desktop\CODE\angular2-webpack-starter\src\app\+detail\detail.routes.ts
Raw Request: loadChildren: './+child-detail#ChildDetailModule'
Replacement: loadChildren: () => System.import('C:\\Users\\shlassaf\\Desktop\\CODE\\angular2-webpack-starter\\src\\app\\+detail\\+child-detail\\index')  .then( (module: any) => module['ChildDetailModule'] )
======================================================================================                                                                                                   69% building modules 790/792 modules 2 active ...rc\app\+detail\+child-detail\index.ts================================== ng-router-loader ==================================
Importer:    C:\Users\shlassaf\Desktop\CODE\angular2-webpack-starter\src\app\+barrel\barrel.routes.ts
Raw Request: loadChildren: './+child-barrel#ChildBarrelModule'
Replacement: loadChildren: () => System.import('C:\\Users\\shlassaf\\Desktop\\CODE\\angular2-webpack-starter\\src\\app\\+barrel\\+child-barrel\\index')  .then( (module: any) => module['ChildBarrelModule'] )
======================================================================================                                                                                                  15756ms building modules
47ms sealing
0ms optimizing
0ms basic module optimization
```

This win32 is also handled in `angular-router-loader` ([see code](https://github.com/brandonroberts/angular-router-loader/blob/master/src/utils.js#L59-L62))

Good for merging now.

@JonnyBGod Is there any change your friend can confirm, just in case?

**Comment by d3viant0ne** [comment]
@shlomiassaf - I spun up a Windows VM, it works fine.

**Comment by pleerock** [comment]
@shlomiassaf I have 500+ lazy routes in my app and I have strange issues - it compiles fine, however when I run the project half of modules cannot be found (runtime error),.

Here is how one of not working routes file look like:

```typescript
export const SymptomConceptPageRoutes: Routes = [
    {
        path: "",
        component: SymptomConceptPage,
        children: [
            {
                path: "",
                pathMatch: "full",
                loadChildren: "modules/medical/symptom/component/concept/symptom-concept-overview-page/SymptomConceptOverviewPageModule#SymptomConceptOverviewPageModule",
            },
            {
                path: "intrasymptomatic-types",
                loadChildren: "modules/medical/intrasymptomatic/component/page/intrasymptomatic-type-list-page/IntrasymptomaticTypeListPageModule#IntrasymptomaticTypeListPageModule"
            },
            {
                path: "publications",
                loadChildren: "modules/content-modules/publication/component/page/publication-list-shared-page/PublicationListSharedPageModule#PublicationListSharedPageModule"
            },
            {
                path: "posts",
                loadChildren: "modules/content-modules/post/component/page/post-list-shared-page/PostListSharedPageModule#PostListSharedPageModule"
            },
            {
                path: "photos",
                loadChildren: "modules/content-modules/photo/component/page/photo-list-shared-page/PhotoListSharedPageModule#PhotoListSharedPageModule"
            },
            {
                path: "videos",
                loadChildren: "modules/content-modules/video/component/page/video-list-shared-page/VideoListSharedPageModule#VideoListSharedPageModule"
            },
            {
                path: "files",
                loadChildren: "modules/content-modules/file/component/page/file-list-shared-page/FileListSharedPageModule#FileListSharedPageModule"
            },
            {
                path: "questions",
                loadChildren: "modules/content-modules/question/component/page/question-list-shared-page/QuestionListSharedPageModule#QuestionListSharedPageModule"
            },
            {
                path: "votes",
                loadChildren: "modules/content-modules/vote/component/page/vote-list-shared-page/VoteListSharedPageModule#VoteListSharedPageModule"
            },
            {
                path: "communities",
                loadChildren: "modules/social/community/component/page/community-list-shared-page/CommunityListSharedPageModule#CommunityListSharedPageModule"
            },
            {
                path: "events",
                loadChildren: "modules/social/event/component/page/event-list-shared-page/EventListSharedPageModule#EventListSharedPageModule"
            },
            {
                path: "information",
                loadChildren: "modules/widget/text-information/component/page/text-information-shared-page/TextInformationSharedPageModule#TextInformationSharedPageModule"
            },
            {
                path: "list",
                loadChildren: "modules/medical/symptom/component/concept/symptom-concept-list-page/SymptomConceptListPageModule#SymptomConceptListPageModule",
            },
            {
                path: "categories",
                loadChildren: "modules/general/category/component/page/category-list-page/CategoryListPageModule#CategoryListPageModule",
            },
            {
                path: "category-types",
                loadChildren: "modules/general/category/component/page/category-type-list-page/CategoryTypeListPageModule#CategoryTypeListPageModule"
            },
            {
                path: "classifications",
                loadChildren: "modules/medical/classification/component/page/classification-concept-list-page/ClassificationConceptListPageModule#ClassificationConceptListPageModule"
            },
        ]
    }
];
```

These routes are inside lazily loaded module.

All modules have correct paths and exist in their directories.
However when I run and navigate throw them I have runtime error looks like this:

`Cannot find module 'modules/content-modules/photo/component/page/photo-list-shared-page/PhotoListSharedPageModule'.`

Again, its runtime error, non compilation/webpack level.

Some of these submodules are working, some of them are not. If I change order of routes in this list, some of them may work, some may broke.

Here is dist output:

```
,children:[{path:"",pathMatch:"full",loadChildren:function(){return n.e(45).then(n.bind(null,1299)).then(function(t){return t.SymptomConceptOverviewPageModule})}},{path:"intrasymptomatic-types",loadChildren:"modules/medical/intrasymptomatic/component/page/intrasymptomatic-type-list-page/IntrasymptomaticTypeListPageModule#IntrasymptomaticTypeListPageModule"},{path:"publications",loadChildren:"modules/content-modules/publication/component/page/publication-list-shared-page/PublicationListSharedPageModule#PublicationListSharedPageModule"},{path:"posts",loadChildren:"modules/content-modules/post/component/page/post-list-shared-page/PostListSharedPageModule#PostListSharedPageModule"},{path:"photos",loadChildren:"modules/content-modules/photo/component/page/photo-list-shared-page/PhotoListSharedPageModule#PhotoListSharedPageModule"},{path:"videos",loadChildren:"modules/content-modules/video/component/page/video-list-shared-page/VideoListSharedPageModule#VideoListSharedPageModule"},{path:"files",loadChildren:"modules/content-modules/file/component/page/file-list-shared-page/FileListSharedPageModule#FileListSharedPageModule"},{path:"questions",loadChildren:"modules/content-modules/question/component/page/question-list-shared-page/QuestionListSharedPageModule#QuestionListSharedPageModule"},{path:"votes",loadChildren:"modules/content-modules/vote/component/page/vote-list-shared-page/VoteListSharedPageModule#VoteListSharedPageModule"},{path:"communities",loadChildren:"modules/social/community/component/page/community-list-shared-page/CommunityListSharedPageModule#CommunityListSharedPageModule"},{path:"events",loadChildren:"modules/social/event/component/page/event-list-shared-page/EventListSharedPageModule#EventListSharedPageModule"},{path:"information",loadChildren:"modules/widget/text-information/component/page/text-information-shared-page/TextInformationSharedPageModule#TextInformationSharedPageModule"},{path:"list",loadChildren:function(){return n.e(33).then(n.bind(null,1297)).then(function(t){return t.SymptomConceptListPageModule})}},{path:"categories",loadChildren:"modules/general/category/component/page/category-list-page/CategoryListPageModule#CategoryListPageModule"},{path:"category-types",loadChildren:"modules/general/category/component/page/category-type-list-page/CategoryTypeListPageModule#CategoryTypeListPageModule"},{path:"classifications",loadChildren:function(){return n.e(39).then(n.bind(null,1127)).then(function(t){return t.ClassificationConceptListPageModule})}}]}]},
```

As you can see some of modules are resolved, some of them are not.

Everything works with previous `'angular-router-loader?loader=system&genDir=compiled/src/app&aot=' + AOT` setting.

I can show you this issue in my project via call if you want.

**Comment by d3viant0ne** [comment]
@pleerock - Probably best opened in https://github.com/shlomiassaf/ng-router-loader

**Comment by shlomiassaf** [comment]
@pleerock please follow the advice from @d3viant0ne and open in the `ng-router-loader` repository.

While you do that please make sure to reference to a GitHub project which I can clone and reproduce myself.

It's doesn't have to be 500+, just something that shows the same error.

Some rule must apply here, on top of my head I can think of 5 different reasons, so I need something to debug.

I will put in effort to solve this for you but you need to help me with reproduction.

**Comment by pglazkov** [comment]
I have the same problem as @pleerock, but on just 4 lazy routes. However, it happens only occasionally. Sometimes it works fine, sometimes it doesn't, without any changes to the source code. I couldn't find any consistent way to reproduce it. Perhaps when you have 500+ lazy routes you could reproduce it more easily.

**Comment by pleerock** [comment]
yeah I just reverted it back to `angular-router-loader` and it works fine now

**Comment by shlomiassaf** [comment]
@pleerock @pglazkov Thank you for your support, I guess.

You probably don't understand the limitations `angular-router-loader` brings with it, if your webpack configuration is simple, yes it will works but it limits all of the goodies webpack provides... 

The problem itself probably relates to fact that `ng-router-loader` is asynchronous so it can use webpack to resolve paths.  
`angular-router-loader` is synchronous and it uses the file system to resolve paths.

Anyone who knows how webpack works realises the problem with the synchronous file system approach.

**Comment by pleerock** [comment]
okay there may be some underlying problems with `angular-router-loader` but at least it works and does what I need. `ng-router-loader` just don't work and reason of bug isn't straightforward and very very strange. I have spend the whole day trying to debug things and try to fix it, and it cost me a lot. Once `ng-router-loader` fixes these issues I'll switch to it.

**Comment by shlomiassaf** [comment]
@pleerock I told you I will put the effort to fix this, just asked for some way to reproduce...

Anyway, I think I know what's the issue.

I won't bother with details but the root cause is not using AST to parse the file, this is something in my todo list... 

The actual problem is probably a race condition on the search/replace algo which is not the smartest but it didnt cause issues (till now) so I didn't refactor it.

I'll post here once fixed.

**Comment by shlomiassaf** [comment]
BTW, i'm still trying to reproduce this...

50 modules by now and still not reproducing, racing bugs are like that...

**Comment by shlomiassaf** [comment]
@pglazkov @pleerock What node version are you using?

I tried with 1000 ngModules and couldn't reproduce it...

@pglazkov I saw that your application (with reference to this issue) is actually using the same setup as the starter... 

So the integration tests in Travis for this repository should fail from time to time....
@d3viant0ne is this the case?

I'm really trying to get to the bottom of this...

**Comment by d3viant0ne** [comment]
Yes, you should be able to restart the job

Travis has issues with parallel execution for free OSS builds

**Comment by d3viant0ne** [comment]
@shlomiassaf - I wonder if we are running into some sort of resource contention / race condition

**Comment by shlomiassaf** [comment]
@pleerock @pglazkov Try version 2.1.0

It uses AST now so it should be a lot better.

I experiments with 2000 lazy modules.

Root has 2 lazy loaded modules.
1 of them (Details) has 1001 lazy modules:
ChildDetail and 1000 .

ChildDetail has 1000 lazy modules inside of it.

You can see it in [this branch](https://github.com/shlomiassaf/angular2-webpack-starter/tree/ng-router-loader-2000-lazy-modules), all with links. 

I couldn't make an AOT production build due to an out of memory exception...
I believe it's because of the compilation dir, the typechecker is also parsing the compilation dir + source dir which is way too much.
If you try running dev, fine.
run AOT, out of memory.
run dev - out of memory - you need to delete the compilation dir.

**Comment by shlomiassaf** [comment]
@d3viant0ne The problem described by @pglazkov @pleerock fit's an issue that might happen with the older implementation (non AST).

The old process used regex with replace to transform the code.
On a module, if multiple lazy routes were set the old process had to take each and replace it.
Before replacing we need to resolve the actual file, this is an async process.
So since each "replace" is **async** I had to store the whole module code in memory and each replace that returned did the actual replace on that in memory "source".

The problem described above shows that some lazy paths were not replaced, which might indicate some race condition but I couldn't reproduce it, I tried 2000 lazy modules, no luck.

What's more confusing is that **async** is not a parallel, every instruction runs alone in the vm so I don't really understand what happened there.

My best guess is that it happens on win machines.

Anyways, I decided to refactor and use AST which is more accurate. I also made sure all changes happens sync.

You can take a look in [this branch](https://github.com/shlomiassaf/angular2-webpack-starter/tree/ng-router-loader-2000-lazy-modules), 2000 lazy modules, works for me.
It won't work on aot... see last issue why I think it doesnt work.

**Comment by JonnyBGod** [comment]
@shlomiassaf I was not happening only on windows machines. I have been having theses errors from time to time. No pattern that I can see just some random modules lost during compile time. It happens in production, dev, aot.

Most times work fine, some time fails, re run and its ok...
Working with macOS, node 6.

Will keep an eye with the 2.1.0 and will let you know if problem persists. Advise to leave the issue open for a week or so as this is really random.

**Comment by pleerock** [comment]
Now I have another problem:

```
ERROR in ./src/modules/human/person/PersonRoutes.ts
Module build failed: TypeError: this.call is not a function
    at Object.loader (/Users/pleerock/www/project/node_modules/index.ts:31:10)
 @ ./src/app/AppRoutes.ts 8:0-68
 @ ./src/app/AppModule.ts
 @ ./src/app/main.browser.ts

ERROR in ./src/modules/general/search/SearchRoutes.ts
Module build failed: TypeError: this.call is not a function
    at Object.loader (/Users/pleerock/www/project/node_modules/index.ts:31:10)
 @ ./src/app/AppRoutes.ts 13:0-70
 @ ./src/app/AppModule.ts
 @ ./src/app/main.browser.ts

..........
```

each routes included in this `AppRoutes` like this:

```typescript
export const AppRoutes: Routes = [
    ...MainHomePageRoutes,
    ...PersonRoutes,
    ...SearchRoutes
```

and each such route file is route definition file like this:

```typescript
export const PersonRoutes: Routes = [
    {
        path: "persons",
        loadChildren: "modules/human/person/component/page/medic-concept-page/MedicConceptPageModule#MedicConceptPageModule",
    },
    {
        path: "person/:id",
        loadChildren: "modules/human/person/component/page/person-view-page/PersonViewPageModule#PersonViewPageModule",
    },
```

each of sub routes has its own lazy routes.

**Comment by JonnyBGod** [comment]
Getting same problem with 2.1.0 as @pleerock with following config:

```
// angular
import { Routes } from '@angular/router';

import { UserLoggedGuard, ProfileExistsGuard } from 'frameworks/app/guards';
import { ProfileResolver } from 'frameworks/app/resolvers';

import { NotFoundComponent } from './components/not-found/not-found.component';

export const ROUTES: Routes = [
  {
    path: '',
    loadChildren: './components/index/index.module#IndexModule'
  },
  {
    path: 'not-found',
    component: NotFoundComponent
  },
  {
    path: 'user',
    loadChildren: './components/user/user.module#UserModule'
  },
  { path: 'login', redirectTo: 'user/login', pathMatch: 'full' },
  { path: 'signin', redirectTo: 'user/login', pathMatch: 'full' },
  { path: 'signup', redirectTo: 'user/signup', pathMatch: 'full' },
  {
    path: 'verify/:token/:userId',
    redirectTo: 'user/verify/:token/:userId',
    pathMatch: 'full'
  },
  { path: 'verify', redirectTo: 'user/verify', pathMatch: 'full' },
  {
    path: 'reset-password/:token/:userId',
    redirectTo: 'user/reset-password/:token/:userId',
    pathMatch: 'full'
  },
  { path: 'recover-account', redirectTo: 'user/recover-account', pathMatch: 'full' },

  // Private
  {
    path: 'organizations',
    loadChildren: './components/organizations/organizations.module#OrganizationsModule',
    canActivate: [ UserLoggedGuard ]
  },
  {
    path: 'settings',
    loadChildren: './components/settings/settings.module#SettingsModule',
    canActivate: [ UserLoggedGuard ]
  },

  // Public
  {
    path: ':id',
    loadChildren: './components/profile/profile.module#ProfileModule',
    canActivate: [ ProfileExistsGuard ],
    resolve: {
      profile: ProfileResolver
    }
  }
];
```

```
ERROR in ./src/app/app.routes.ts
Module build failed: TypeError: this.call is not a function
    at Object.loader (/Users/joaoribeiro/Documents/Projects/seeds/web-seed/node_modules/index.ts:31:10)
 @ ./src/app/app.routerModule.ts 2:0-38
 @ ./src/app/app.advance.ts
 @ ./src/app/app.module.ts
 @ ./src/app/index.ts
 @ ./src/main.browser.ts
 @ multi main
```

**Comment by JonnyBGod** [comment]
@pleerock solved it by changing the loaders order as in https://github.com/shlomiassaf/angular2-webpack-starter/blob/ng-router-loader-2000-lazy-modules/config/webpack.common.js#L102-L117

**Comment by shlomiassaf** [comment]
my-router-loader uses JavaScript AST analysis to find the loadChildren string.

This is what it must come after the awesome typescript loader is done...

**Comment by shlomiassaf** [comment]
@JonnyBGod what about the random fails?

**Comment by JonnyBGod** [comment]
All good for now. But give it a few more days.

**Comment by pleerock** [comment]
thanks @JonnyBGold looks like its working right now

**Comment by pglazkov** [comment]
@shlomiassaf Sorry for a late reply. Here are the answers to your questions:

- Node Version: 6.9.4
- Yes, I'm using Windows

I would be glad to help with reproducing the issue, but, as I said before, I couldn't find any ways to reproduce it consistently. Indeed, it looks like some kind of race condition. 

I will try 2.1.0 for a few days and will let you know if it fixes the problem.

**Comment by pglazkov** [comment]
@shlomiassaf As for why the integration tests in Travis for this repository don't fail from time to time, maybe that's because the e2e tests don't cover the lazy routes? 
As far as I can see, the e2e test just checks some stuff on the home page, but when the lazy routes are not compiled correctly the home page still loads and works fine, but there are runtime JS errors when routes are being pre-loaded.

**Comment by shlomiassaf** [comment]
@pglazkov Your project had 4 lazy routes and failed sometimes, since @pleerock had 500+ lazy routes and he approved the fix to be working I consider this issue closed...

Let me know if otherwise.

**Comment by JonnyBGod** [comment]
@shlomiassaf All good for me until now. Apparently it is fixed.
